tracing: add generic infrastructure for collection of tracing data

wentongwu opened the pull request:
1. add generic infrastructure for collection of tracing data.

2. add samples for tracing framework with the help of ctf format.

3. add script to receive the ctf data which can be decoded by trace compass.

4. add usb backend.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

mped-oticon commented on subsys/debug/tracing/include/tracing_backend.h:
```diff
@@ -0,0 +1,139 @@
+ * @}
+ */
+
+#define TRACING_BACKEND_UART_NAME "tracing_backend_uart"
```
This header is generic, so why have it tainted by UART stuff?

wentongwu replied:
I will use the name of backend to get the working backend instance, I think this is the most suitable file where we place them.

mped-oticon commented on subsys/debug/tracing/include/tracing_backend.h:
```diff
@@ -0,0 +1,139 @@
+	{                                                                \
+		.name = STRINGIFY(_name),                                \
+		.api = &_api,                                            \
+		.cb = &_name##_cb                                        \
```
Would it be possible to avoid _pointing_ to the control block? It doesn't change as far as I can tell, so it could be inplace.

mped-oticon commented on subsys/debug/tracing/include/tracing_backend.h:
```diff
@@ -0,0 +1,139 @@
+		struct tracing_packet *packet)
+{
+	if (backend && backend->api && backend->api->output) {
+		backend->api->output(backend, packet);
```
Again, is there a need for pointing to api? It shouldn't change, so why permit it (checks and derefs cost code and cycles). Rather prefer `backend.api.output(backend, packet)`.

wentongwu replied:
@mped-oticon thanks, updated.

mped-oticon commented on subsys/debug/tracing/ctf/CMakeLists.txt:
```diff
@@ -12,3 +12,4 @@ zephyr_library_include_directories(
 zephyr_include_directories(.)
 
 add_subdirectory_ifdef(CONFIG_TRACING_CTF_BOTTOM_POSIX bottoms/posix)
+add_subdirectory_ifdef(CONFIG_TRACING_CTF_BOTTOM_GENERIC bottoms/generic)
```
Cool, thanks for adding the first non-posix CTF backend.

mped-oticon commented on samples/subsys/debug/tracing/src/main.c:
```diff
@@ -0,0 +1,29 @@
+		printk("Error: Console device name is not USB ACM\n");
+
+		return;
+	}
```
Given that 1) the while-loop below is the demo to be traced and 2) the location of this main.c -- then it would be nice if this demo app wasn't specific for any particular tracing backend.
This code might as well be part of the compiled-in backend's init(), no?

wentongwu replied:
@mped-oticon updated. Thanks

mped-oticon commented on CODEOWNERS:
```diff
@@ -333,6 +333,7 @@
 /scripts/gen_syscalls.py                  @andrewboie
 /scripts/process_gperf.py                 @andrewboie
 /scripts/gen_relocate_app.py              @wentongwu
+/scripts/receive_ctf.py                   @wentongwu
```
I'd propose to change the name: It actually connects to serial, rather than doing anything with CTF. It just then _happens_ to be that what it receives is a CTF stream.

`receive_serial.py`

Oticon plans on contributing scripts to parse actual CTF streams (agnostic of IO delivery mechanism) for improved visualization in tools like TraceCompass and/or KernelShark. Hence to avoid confusion and name collision, this name change.

wentongwu replied:
@mped-oticon thanks, updated, I like your mentioned tools, please include me for review, thanks.

mped-oticon commented on subsys/debug/tracing/CMakeLists.txt:
```diff
@@ -1,27 +1,37 @@
+  CONFIG_TRACING
+  tracing_core.c
+  tracing_format.c
+  tracing_packet.c
```
These sources should only be included under `CONFIG_TRACING_CTF_BOTTOM_GENERIC`.
E.g. Oticon has a proprietary CTF bottom (so CONFIG_TRACING is on) but we do not need this new buffered GENERIC bottom.

mped-oticon replied:
Still unresolved. Please specialize from CONFIG_TRACING to something that actually depends on this, e.g. CONFIG_TRACING_GENERIC (or CONFIG_TRACING_BUFFERED, preferrable IMHO).

mped-oticon reviewed (changes requested):
A welcome change. Good to see another CTF backend. The first public non-posix ctf bottom :+1: 

This PR is not a Framework: This PR plugs into the existing CTF bottom API (which itself is plugged into the Zephyr trace API).

This PR is not 'Generic': It provides for different hardware backends, yes, and comes with a default one (CDC UART). But the common denominator for this PR's CTF bottom is that it is **buffered**.

So I propose to change 'Generic' to 'Buffered'.

mped-oticon commented:
To clarify, the situation hierarchically:

```
Zephyr Trace API
 |-> Segger SystemView
 \-> CTF top. Can compile-select between several bottoms:
      |-> IO bottom: posix (using fwrite)
      |-> IO bottom: buffered (this PR). This buffered CTF bottom is Generic in the sense that further IO bottoms can be implemented (i.e. implementing tracing_backend_api)
      |    \-> tracing_backend_uart. 
      \-> IO bottom: Oticon proprietary (memory mapped, unbuffered)
```

I'd propose changing this PR's name to: "tracing: Add new CTF bottom: Buffered and extensible, optionally formatted".

wentongwu commented:
> A welcome change. Good to see another CTF backend. The first public non-posix ctf bottom +1
> 
> This PR is not a Framework: This PR plugs into the existing CTF bottom API (which itself is plugged into the Zephyr trace API).
> 
> This PR is not 'Generic': It provides for different hardware backends, yes, and comes with a default one (CDC UART). But the common denominator for this PR's CTF bottom is that it is **buffered**.
> 
> So I propose to change 'Generic' to 'Buffered'.

Thanks for the review. @mped-oticon 
This tracing framework is not designed for CTF using only, it's used for collecting many kinds of tracing format(front-end), I think CTF is one kind of front-ends, I use the existing CTF format to verify this infrastructure can work well. Maybe later we can add other front-ends on top of it.

For the unbuffered concerns, actually I also consider it before coding(I called it synchronous method), but I'm not sure if it will impact the real time of the system if back-end is slow, so I post the code first to review, if synchronous method is also needed, I will add it also. 

And a USB back-end is also under developing.

fnde-ot reviewed (commented):
At first glance this looks like a great contribution! I also have a few doubts to how this should be plugged in.

I understand this is meant to be a generic buffered backend to be used with multiple frontends, however CTF is itself meant to be a generic frontend as it could be made do define -almost- any output format. Shouldn't this then be made a CTF bottom as @mped-oticon suggests?

Also, could this also be used as a LOG or PRINTK backend, instead of being limited to TRACING?

wentongwu commented:
> Also, could this also be used as a LOG or PRINTK backend, instead of being limited to TRACING?

@fnde-ot sure, it can be used as printk or log backend, see `tracing_format_string`

fnde-ot commented:
> @fnde-ot sure, it can be used as printk or log backend, see `tracing_format_string`

Awesome! Then I would suggest to place this in a new subsystem or as a library so it can be used by more than just tracing.

wentongwu commented:
> > @fnde-ot sure, it can be used as printk or log backend, see `tracing_format_string`
> 
> Awesome! Then I would suggest to place this in a new subsystem or as a library so it can be used by more than just tracing.

@fnde-ot if anything wrong, I apologize for that. I know there is huge improvements needed for the current code, I just want to focus on tracing here, yes, I just try it on CTF on which I have not so much experience, but also I have plan to try percepio tracealyzer, not sure so far CTF can cover the format of tracealyzer and others', so that's the reason why I want to put the code in same level path as ctf.

fnde-ot commented:
> @fnde-ot if anything wrong, I apologize for that. I know there is huge improvements needed for the current code, I just want to focus on tracing here, yes, I just try it on CTF on which I have not so much experience, but also I have plan to try percepio tracealyzer, not sure so far CTF can cover the format of tracealyzer and others', so that's the reason why I want to put the code in same level path as ctf.

No need to apologize, this is a great contribution and maybe it could stay in the tracing subsystem for now as a CTF bottom (we should still definitely consider it for other subsystems requiring buffered serial output).

As far as percepio tracealyzer goes, I believe the plan was to use CTF output format for that tool, as well as others (tracecompass, kernelshark, ...). See the TSC discussion topic here:
https://lists.zephyrproject.org/g/tsc/topic/trace_visualization/31546422

Maybe @nashif or someone from the TSC should take a look at this PR.

wentongwu commented:
> > @fnde-ot if anything wrong, I apologize for that. I know there is huge improvements needed for the current code, I just want to focus on tracing here, yes, I just try it on CTF on which I have not so much experience, but also I have plan to try percepio tracealyzer, not sure so far CTF can cover the format of tracealyzer and others', so that's the reason why I want to put the code in same level path as ctf.
> 
> No need to apologize, this is a great contribution and maybe it could stay in the tracing subsystem for now as a CTF bottom (we should still definitely consider it for other subsystems requiring buffered serial output).
> 
> As far as percepio tracealyzer goes, I believe the plan was to use CTF output format for that tool, as well as others (tracecompass, kernelshark, ...). See the TSC discussion topic here:
> [lists.zephyrproject.org/g/tsc/topic/trace_visualization/31546422](https://lists.zephyrproject.org/g/tsc/topic/trace_visualization/31546422)
> 
> Maybe @nashif or someone from the TSC should take a look at this PR.

@fnde-ot @mped-oticon if the decision to using CTF to support all kinds of tracing tools has made, I'm fine with your proposals and I will change the place where we put the code, but could you please help review the code again?

wentongwu commented:
and the CI failed because of https://github.com/zephyrproject-rtos/zephyr/issues/21048, not related to this change

wentongwu commented:
@fnde-ot @mped-oticon could you please review this PR again? thanks. I will submit USB backend after this PR is merged.

mped-oticon commented:
> @fnde-ot @mped-oticon could you please review this PR again? thanks. I will submit USB backend after this PR is merged.

Yep, will do -- but will first happen for me early next week. X-mas parties are imposing themselves ;-)

nashif commented on include/debug/tracing_format.h:
```diff
@@ -0,0 +1,87 @@
+/*
```
files added in include/debug/ are generic and not related to ctf, split them out into their own commit.

nashif commented on include/debug/tracing_core.h:
```diff
@@ -0,0 +1,54 @@
+/*
```
files added in include/debug/ are generic and not related to ctf, split them out into their own commit.

nashif commented on include/linker/common-rom.ld:
```diff
@@ -160,3 +160,10 @@
 		KEEP(*(SORT_BY_NAME("._cfb_font.*")))
```
This file is generic and not related to ctf, split them out into their own commit.

nashif commented on subsys/debug/Kconfig:
```diff
@@ -3,6 +3,8 @@
 # Copyright (c) 2015 Wind River Systems, Inc.
```
split the changes in this file into 2 parts, generic tracing backend, and CTF related changes.

nashif reviewed (changes requested):
Here a first batch of comments, more to come later today.

This should not be seen as CTF bottom only, the PR might uses CTF as the tracing format right now, because this one is supported in Zephyr and is available, but we should be able to send any format, also tracing via printk, sys-t or any other mean. For example we could implement a tracing backend using just printk that can be sent to the host using one of the backends (uart, usb, networking, filesystem, etc.)

Already in this review I can see conflicting requests from @mped-oticon  and @fnde-ot regarding what this should ultimately support (generic vs ctf only).

nashif commented on scripts/receive_uart.py:
```diff
@@ -0,0 +1,52 @@
+    ser.close()
+
+if __name__=="__main__":
+    main()
```
Maybe:
```
if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('Data capture interrupted, data saved into {}'.format(args.output))
        sys.exit(0)

```

nashif commented on samples/subsys/debug/tracing/src/main.c:
```diff
@@ -0,0 +1,19 @@
+	while (1) {
+		printk("%c\n", (count++) % 2 ? 'A' : 'B');
+
+		k_sleep(K_SECONDS(2));
```
Not much to trace here, you probably should reuse 

`samples/synchronization/src/main.c` which has 2 threads and a semaphore :)

nashif reviewed (changes requested):
more comments

nashif commented:
> To clarify, the situation hierarchically:
> 
> ```
> Zephyr Trace API
>  |-> Segger SystemView
>  \-> CTF top. Can compile-select between several bottoms:
>       |-> IO bottom: posix (using fwrite)
>       |-> IO bottom: buffered (this PR). This buffered CTF bottom is Generic in the sense that further IO bottoms can be implemented (i.e. implementing tracing_backend_api)
>       |    \-> tracing_backend_uart. 
>       \-> IO bottom: Oticon proprietary (memory mapped, unbuffered)
> ```
> 
> I'd propose changing this PR's name to: "tracing: Add new CTF bottom: Buffered and extensible, optionally formatted"

Hmm, unless I am missing something, this hierarchy assumes CTF is the center of the world and every backend we need to support needs to be a CTF bottom? 

The intention of this PR was/is to provide a common infrastructure for pushing tracing data to the host, so instead of putting all of this under CTF I would suggest to move the current bottoms (in this case posix to be another backend of this generic infrastructure and not make new backends CTF specific.

Why something as generic as `tracing_backend_uart.c` or `tracing_core.c` need to be under CTF and depend on CTF? 

I propose to make the native_posix fwrite so called bottom a backend of this new infrastructure and avoid code duplication and make any existing backends available to any new tracing formats we might add.

wentongwu committed:
tracing: add generic infrastructure for collection of tracing data

Add generic infrastructure for collection of tracing data. And it can
handle data format message and string format message, the handled
packet will go through tracing thread, and then the tracing backend
will output the tracing packet to host.

Currently we enable UART and USB backend following tracing_backend_api.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tracing: ctf: add bottom for ctf with the help generic infrastructure

Add bottom for ctf data format with the help of generic infrastructure.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
scripts: add script to capture tracing data with UART backend

Add script to capture tracing data with UART backend.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
scripts: add script for usb backend to receive tracing data

Add script for usb backend to receive tracing data.
The script, trace_capture_usb.py, is based on pyusb, so
install it correctly before using the script.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tracing: remove the ISR tracing hook in ISR wrapper.

Currently the ISR tracing hook is put at the beginning and end of ISR
wrapper, so every ISR will be traced, when there is ISR needed in the
tracing path(especially tracing back-end, every output of tracing
back-end will be traced once), it will cause the tracing buffer easily
be exhausted, also it will increase system latency, so remove the
tracing hook at isr_wrapper, just tracing system timer though.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

dbkinder commented on samples/synchronization/README.rst:
```diff
@@ -44,3 +49,33 @@ Sample Output
+*******************************
+Build image with "cmake -DBOARD=sam_e70_xplained -DCONF_FILE=prj_uart.conf .."
+
+After the board runnig a while, running script trace_capture_uart.py with
```
How about:
```
After the application has run for a while, run the
``trace_capture_uart.py`` script on your host::

   sudo python3 trace_capture_uart.py -d /dev/ttyACM1 -b 115200 -o log_1

The sam_e70_xplained's serial port is ``/dev/ttyACM1``
and is used as the tracing backend. The script's output file
``log_1`` and metadata under ``zephyr/subsys/debug/tracing/ctf/tsdl``
can be used to serve Tracecompass.
```

Where is Tracecompass described?

wentongwu replied:
@dbkinder many thanks, updated, also the Tracecompass is here https://www.eclipse.org/tracecompass/

dbkinder commented on samples/synchronization/README.rst:
```diff
@@ -44,3 +49,33 @@ Sample Output
+
+Usage for UART tracing backend
+*******************************
+Build image with "cmake -DBOARD=sam_e70_xplained -DCONF_FILE=prj_uart.conf .."
```
Change to:
```
Build a UART-tracing image with::

   cmake -DBOARD=sam_e70_xplained -DCONF_FILE=prj_uart.conf ..
```
(Note that the two colons ``::`` at the end of the line indicate the following intented text is a code-block.)

wentongwu replied:
thanks

dbkinder commented on samples/synchronization/README.rst:
```diff
@@ -44,3 +49,33 @@ Sample Output
+
+Usage for USB tracing backend
+******************************
+Build image with "cmake -DBOARD=sam_e70_xplained -DCONF_FILE=prj_usb.conf .."
```
Change to:
```
Build a USB-tracing image with::

   cmake -DBOARD=sam_e70_xplained -DCONF_FILE=prj_usb.conf ..
```

dbkinder commented on samples/synchronization/README.rst:
```diff
@@ -44,3 +49,33 @@ Sample Output
+******************************
+Build image with "cmake -DBOARD=sam_e70_xplained -DCONF_FILE=prj_usb.conf .."
+
+After serial console has stable output like below, connect board usb port to
```
Change to:
```
After the serial console has stable output like this::

   threadA: Hello World!
   threadB: Hello World!
   threadA: Hello World!
   threadB: Hello World!

connect the board's USB port to the host device and
run the ``trace_capture_usb.py`` script on the host::

   sudo python3 trace_capture_usb.py -v 0x2FE9 -p 0x100 -o log_2

The script's output file ``log_2`` and metadata under
``zephyr/subsys/debug/tracing/ctf/tsdl`` can
be used to serve Tracecompass.

```

wentongwu replied:
@dbkinder  many thanks, updated!

nashif commented on samples/synchronization/README.rst:
```diff
@@ -12,6 +12,11 @@ and use sleep requests and semaphores to control the rate at which messages
 are generated. This demonstrates that kernel scheduling, communication,
 and timing are operating correctly.
 
+And this application can also be used to demonstrate the tracing feature. The
```
sorry if I confused you, I did not mean for you to reuse the existing sample, I was saying that you could reuse the code. I prefer to leave this sample as simple as possible, anything with tracing can be added in a tracing specific sample that can be extended to show tracing specific features.

You could use this one if you wish: `tests/subsys/debug/tracing`

wentongwu commented on samples/synchronization/README.rst:
```diff
@@ -12,6 +12,11 @@ and use sleep requests and semaphores to control the rate at which messages
 are generated. This demonstrates that kernel scheduling, communication,
 and timing are operating correctly.
 
+And this application can also be used to demonstrate the tracing feature. The
```
Thanks

wentongwu committed:
tests: add tracing test based on the existing testing code

Add tracing test based on the existing testing code.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

dbkinder reviewed (commented):
There are no doc changes now in this PR.  Should there be an update to https://docs.zephyrproject.org/latest/guides/tracing/index.html ?

mped-oticon commented on scripts/tracing/trace_capture_usb.py:
```diff
@@ -0,0 +1,99 @@
+            data = []
+            data = read_endpoint.read(read_endpoint.wMaxPacketSize, 100000)
+            for c in data:
+                file_desc.write(chr(c).encode('latin1'))
```
premature optimization thoughts: character-by-character encoding could be a bottleneck for a high-throughput trace stream. I'd investigate if it could be done on at the block level (if nessecary).

mped-oticon commented on scripts/tracing/trace_capture_usb.py:
```diff
@@ -0,0 +1,99 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
+Script to capture tracing data with USB backend.
```
Script LGTM. 
s/USB backend/tracing USB CDC backend/

mped-oticon commented on subsys/debug/Kconfig:
```diff
@@ -350,6 +350,12 @@ config TRACING_CTF_BOTTOM_POSIX
 	  Enable POSIX backend for CTF tracing. It will output the CTF stream to a
 	  file using fwrite.
 
+config TRACING_CTF_BOTTOM_GENERIC
```
I still think it should be called "buffered" rather than "generic", as that is what it is.

mped-oticon commented on subsys/debug/tracing/CMakeLists.txt:
```diff
@@ -1,27 +1,42 @@
+  )
+
+zephyr_sources_ifdef(
+  CONFIG_TRACING_BACKEND_USB
```
OK. Prefer: CONFIG_TRACING_(GENERIC|BUFFERED)_USB

mped-oticon reviewed (changes requested):
Generally looks OK, still I have an issue with the name "Generic". It sounds confusing and vague too. 'Buffered' would be more clear, and delimit architectural choice of backend implementions -- unbuffered/synchronous ones versus buffered/assync ones.

mped-oticon commented:
@nashif 

> Hmm, unless I am missing something, this hierarchy assumes CTF is the center of the world and every backend we need to support needs to be a CTF bottom?

If you looked at my ASCII tree, it should be clear that isn't the case. Namely, Segger sits next to CTF. So CTF is not "center of the world" (but IMHO it should be). Why would we not consolidate around an {open, efficient, ...} standard like CTF, with a growing ecosystem of tools around it. 


> The intention of this PR was/is to provide a common infrastructure for pushing tracing data to the host, so instead of putting all of this under CTF I would suggest to move the current bottoms (in this case posix to be another backend of this generic infrastructure and not make new backends CTF specific.

Disagree. Explanation:

0. Oticon is already pushing data to the host. No buffering needed. 
1. This PR is **buffered**. We at Oticon do not want that. We're happy with our _prioprietary_  synchronous memory mapped CTF backend. What this PR is doing is essentially replicating what was done with the logging framework; buffering stuff up until a slow UART has some air to breathe. Not efficient (more cycles wasted per trace), complex (overflow policy anyone?), and has dubious timestamping. 
2. The current CTF infrastructure in Zephyr is already sufficient and efficient, why do you want to wrap it again? The map macro and much of this PR has actually been copied already from the CTF backend. 
3. The current CTF infrastructure in Zephyr is backend **agnostic** and goes to great lengths to permit any kind of backend. This PR being a good example, even though it replicates some existing macros and idioms.
4. It does make sense to wrap it again, if one wants to **buffer**. But not everybody wants that.

 
> Why something as generic as `tracing_backend_uart.c` or `tracing_core.c` need to be under CTF and depend on CTF?

Because it would be nice to consolidate on one format-to-rule-them-all? I.e. CTF. And retain one TSDL for that definition along with one CTF top that defines what gets emitted.
But I don't want them there if merely CTF is kconfig chosen, I want them there if CTF_BUFFERED is there.

Secondly, `tracing_core.c` does sound generic, except it _isn't_. It assumes specifically a buffered backend. Seems like POSIX fwrite CTF bottom does fine without it.

 
> I propose to make the native_posix fwrite so called bottom a backend of this new infrastructure and avoid code duplication and make any existing backends available to any new tracing formats we might add.

That would mean buffered tracing becomes unavoidable. No thanks.
The current CTF implementation has very little overhead, almost none if you implement it correctly -- compared with this buffered PR.

I don't want to allocate a whole new thread or increase stack size for doing my very efficient synchronous tracing emit. As a result, I regard this PR as imposing itself a bit too much, making assumptions while calling itself 'generic'. Frankly, overruling the thoughts that led to the generality already exhibited by CTF; you can read more about these thoughts in the documentation: https://docs.zephyrproject.org/latest/guides/tracing/ctf.html#i-o-taxonomy

wentongwu commented:
> 1. Oticon is already pushing data to the host. No buffering needed.
> 2. This PR is **buffered**. We at Oticon do not want that. We're happy with our _prioprietary_  synchronous memory mapped CTF backend. What this PR is doing is essentially replicating what was done with the logging framework; buffering stuff up until a slow UART has some air to breathe. Not efficient (more cycles wasted per trace), complex (overflow policy anyone?), and has dubious timestamping.
> 3. The current CTF infrastructure in Zephyr is already sufficient and efficient, why do you want to wrap it again? The map macro and much of this PR has actually been copied already from the CTF backend.

@mped-oticon For the MAP marco, yes, I get the idea from you (subsys/debug/tracing/ctf/bottoms/posix/ctf_bottom.h) when try CTF on this infrastructure
, sorry for wrap it, I will put THE MARCO under subsys/debug/tracing/ctf/bottoms/generic/
, but what do you mean for `and much of this PR has actually been copied already from the CTF backend` ?

wentongwu commented:
> That would mean buffered tracing becomes unavoidable. No thanks.
> The current CTF implementation has very little overhead, almost none if you implement it correctly -- compared with this buffered PR.

@mped-oticon I agree it's buffered. But not every platform has RTT or your mentioned `prioprietary synchronous memory mapped CTF backend`, instead most platforms have UART and USB, etc, and I would like to see your very little/none overhead implementation with current code for UART and USB.

mped-oticon commented:
> > That would mean buffered tracing becomes unavoidable. No thanks.
> > The current CTF implementation has very little overhead, almost none if you implement it correctly -- compared with this buffered PR.
> 
> @mped-oticon I agree it's buffered. But not every platform has RTT or your mentioned `prioprietary synchronous memory mapped CTF backend`, instead most platforms have UART and USB, etc, and I would like to see your very little/none overhead implementation with current code for UART and USB.

No offense meant. I like this PR, thanks for it. It is definitely needed for UART etc.
I just don't want the roles swapped in the future as @nashif indicates, as that would be a shame.
This PR will operate fine being a regular CTF bottom.

mped-oticon commented:
> @mped-oticon For the MAP marco, yes, I get the idea from you (subsys/debug/tracing/ctf/bottoms/posix/ctf_bottom.h) when try CTF on this infrastructure
> , sorry for wrap it, I will put THE MARCO under subsys/debug/tracing/ctf/bottoms/generic/
> , but what do you mean for `and much of this PR has actually been copied already from the CTF backend` ?

I'm fine this PR. Copying is also fine. I was just saying some infrastructure could perhaps be the same, and some copying could have been avoided.

nashif commented:
> If you looked at my ASCII tree, it should be clear that isn't the case. Namely, Segger sits next to CTF. So CTF is not "center of the world" (but IMHO it should be). Why would we not consolidate around an {open, efficient, ...} standard like CTF, with a growing ecosystem of tools around it.

I am referring to the bottom half and the transport, not the for payload format, why would want to we have all the backends (which are generic and should be payload agnostic) under CTF?
I should be able to create a dummy backend that just uses printf and push the data via UART or whatever, why do I have to go via  the CTF IO bottom ?

> * Oticon is already pushing data to the host. No buffering needed.
> 
> * This PR is **buffered**. We at Oticon do not want that. We're happy with our _prioprietary_  synchronous memory mapped CTF backend. What this PR is doing is essentially replicating what was done with the logging framework; buffering stuff up until a slow UART has some air to breathe. Not efficient (more cycles wasted per trace), complex (overflow policy anyone?), and has dubious timestamping.

This PR was not trying to replace the existing non-buffered transport, it was merely adding additional backends. We are aware that the buffered approach is not effecient, but on some systems this is the only way possible and is provided as the last resort when no other more effecient backends are available. Where applicable, more efficient backends can be added.

>     5\. It does make sense to wrap it again, if one wants to **buffer**. But not everybody wants that.

and if you do not want buffering, you are free not to use it and stick with you own implementation, I am not sure what the issue is. 

> That would mean buffered tracing becomes unavoidable. No thanks.
> The current CTF implementation has very little overhead, almost none if you implement it correctly -- compared with this buffered PR.

Why would a native posix fwrite be buffered? This should something that is defined by the backend, you could have backends that would need buffering, others that do not, i.e. write directly to some memory or whatever.

to sum it up, CTF is a tracing format, not a transport, there are other formats, some are less efficient and some maybe more efficient that CTF, there are propitiatory and open standards and making the CTF the one and only is a NO go. Thats why I think it makes lots of sense to seperate transport from format and support different types of both formats and transports, both buffered and non-buffered.
